### PR TITLE
Enhance logging for CallbackHandler and Zkclient

### LINKED ISSUE #1352: Improve CallbackHandler and ZkClient logging
CallbackHandlder and ZkClient methods are invoked over multiple threads. It can be hard to correlate one object's execution threads over multiple java threads. This poses a serious issue in production log examination.

We propose to add unique id to each methods' logging to help the correlation and also track the object's life cycle.

### PR DESCRIPTION
**Pull request by kaisun2000**
### Issues

- [x] My PR addresses the following Helix issues and references them in the PR description:

resolve #1352

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

    CallbackHandlder and ZkClient methods are invoked over multiple threads.
    It can be hard to correlate one object's execution threads over multiple
    java threads. This poses a serious issue in production log examination.
    
    We propose to add unique id to each methods' logging to help the
    correlation and also track the object's life cycle.

### Tests

- [x] The following tests are written for this issue:

None

- [x] The following is the result of the "mvn test" command on the appropriate module:

running

### Documentation (Optional)

- In case of new functionality, my PR adds documentation in the following wiki page:

(Link the GitHub wiki you added)

### Commits

- My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

File: helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java [hunk 1]
```diff
@@ -660,24 +666,31 @@ private void subscribeForChanges(NotificationContext.Type callbackType, String p
...
 
   public EventType[] getEventTypes() {
     return (EventType[]) _eventTypes.toArray();
   }
 
+  @Override
+  public String toString() {
+    StringBuilder stringBuilder = new StringBuilder();
+    stringBuilder.append(_uid);
+    return stringBuilder.toString();
```
[the remark this answers is not in the document]
Comment (kaisun2000) [reply helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java:691]: I think Lei mentioned in the review meeting not to hard code the UID. How about let me use a function called getUID()?

[the remark this answers is not in the document]
Comment (jiajunwang) [reply helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java:691]: I agree to what Huizhi mentioned, please include the basic information. If only uid is needed, then just refer to the _uid field in the code. No need to override toString().

File: zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java [hunk 2]
```diff
@@ -204,6 +204,13 @@ public void recordPathStat(Stat stat, OptionalLong notificationTime) {
     }
   }
 
+  @Override
+  public String toString() {
```
Comment (jiajunwang) [inline zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:208]: Same thing here, it seems that directly refer to _uid is enough.

File: zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java [hunk 3]
```diff
@@ -220,7 +227,7 @@ protected ZkClient(IZkConnection zkConnection, int connectionTimeout, long opera
 
     _asyncCallRetryThread = new ZkAsyncRetryThread(zkConnection.getServers());
     _asyncCallRetryThread.start();
-    LOG.debug("ZkClient created with _uid {}, _asyncCallRetryThread id {}", _uid, _asyncCallRetryThread.getId());
+    LOG.debug("ZkClient created with _uid {}, _asyncCallRetryThread id {}", this.toString(), _asyncCallRetryThread.getId());
```
Comment (jiajunwang) [inline zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:230]: nit "_uid" => "uid"

File: zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java [hunk 4]
```diff
@@ -204,6 +204,13 @@ public void recordPathStat(Stat stat, OptionalLong notificationTime) {
     }
   }
 
+  @Override
+  public String toString() {
+    StringBuilder stringBuilder = new StringBuilder();
+    stringBuilder.append(_uid);
```
Comment (jiajunwang) [inline zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:210]: I just notice that _uid here was not made final field, could you please fix it? It should be a final field.

File: zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java (hunk 3 above)
Comment (kaisun2000) [reply zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java]: changed,

File: zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java (hunk 4 above)
Comment (kaisun2000) [reply zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java]: changed to final

File: zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java (hunk 2 above)
Comment (kaisun2000) [reply zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java]: changed.

File: helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (kaisun2000) [reply helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java]: changed .

File: zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java [hunk 5]
```diff
@@ -2228,7 +2233,7 @@ public void close() throws ZkInterruptedException {
       _eventThread.interrupt();
       _eventThread.join(2000);
       if (isManagingZkConnection()) {
-        LOG.info("Closing zkclient: " + ((ZkConnection) connection).getZookeeper());
+        LOG.info("zkclient{}, Closing zkclient zk: {} ", this._uid, ((ZkConnection) connection).getZookeeper());
```
Comment (pkuwm) [inline zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:2236]: It'd be great to make these similar messages easier to read, like `Closing zkclient. uid={}, zk={}`?

File: helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java [hunk 6]
```diff
@@ -105,7 +106,9 @@
 @PreFetchChangedData(enabled = false)
 public class CallbackHandler implements IZkChildListener, IZkDataListener {
   private static Logger logger = LoggerFactory.getLogger(CallbackHandler.class);
+  private static AtomicLong CB_UID = new AtomicLong();
```
Comment (pkuwm) [inline helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java:109]: `static final`?
And also give it a more comprehensive name, like `CALLBACK_HANDLER_UID` or `NEXT_ID`?

File: helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java [hunk 7]
```diff
@@ -28,6 +28,7 @@
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
```
Comment (pkuwm) [inline helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java:31]: This is not used. Remove it?

File: helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java (hunk 6 above)
Comment (kaisun2000) [reply helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java:109]: changed.

**Review by pkuwm — APPROVED** [review]
LGTM.

File: helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java [hunk 8]
```diff
@@ -105,7 +105,9 @@
 @PreFetchChangedData(enabled = false)
 public class CallbackHandler implements IZkChildListener, IZkDataListener {
   private static Logger logger = LoggerFactory.getLogger(CallbackHandler.class);
+  private static final AtomicLong CALLBACK_HANDLER_UI = new AtomicLong();
```
Comment (pkuwm) [inline helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java:108]: @kaisun2000 This one? `CALLBACK_HANDLER_UI` -> `CALLBACK_HANDLER_UID`

Comment (kaisun2000) [reply helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java]: good catch. fixed.

**Comment by kaisun2000** [comment]
This pull request is approved, please help to merge

>CallbackHandlder and ZkClient methods are invoked over multiple threads.
It can be hard to correlate one object's execution thread over multiple
java threads. This poses a serious issue in production log examination.

We propose to add unique id to each methods' logging to help the
correlation and also track the object's life cycle.

File: helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java [hunk 9]
```diff
@@ -354,10 +359,10 @@ public void enqueueTask(NotificationContext changeContext) throws Exception {
     // async mode only applicable to CALLBACK from ZK, During INIT and FINALIZE invoke the
     // callback's immediately.
     if (_batchModeEnabled && changeContext.getType() == NotificationContext.Type.CALLBACK) {
-      logger.debug("Enqueuing callback");
+      logger.debug("Callbackhandler {}, Enqueuing callback", this._uid );
       if (!isReady()) {
-        logger.info("CallbackHandler is not ready, ignore change callback from path: {}, for "
-            + "listener: {}", _path, _listener);
+        logger.info("CallbackHandler {} is not ready, ignore change callback from path: {}, for "
+            + "listener: {}", this._uid, _path, _listener);
```
Comment (jiajunwang) [inline helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java:365]: nit, "this." in "this._uid" is not necessary, right?

**Review by jiajunwang — APPROVED** [review]

File: helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java (hunk 9 above)
Comment (kaisun2000) [reply helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java]: removed all this.

**Pull request merged**